src/main/java/com/maxmind/db/SingleBuffer.java

```diff
@@ -0,0 +1,174 @@
+                "SingleBuffer cannot exceed Integer.MAX_VALUE capacity"
+            );
+        }
+        this.buffer = ByteBuffer.allocate((int) capacity);
+    }
+
+    /**
+     * Creates a new {@code SingleBuffer} wrapping the given {@link ByteBuffer}.
+     *
+     * @param buffer the underlying buffer
+     */
+    private SingleBuffer(ByteBuffer buffer) {
+        this.buffer = buffer;
+    }
+
+    /**
+     * Validates and converts a long index to an int for use with
+     * {@link ByteBuffer}.
+     *
+     * @param index the index to check
+     * @return the index as an int
+     * @throws IndexOutOfBoundsException if the index is negative or exceeds
+     *                                   {@link Integer#MAX_VALUE}
+     */
+    private int checkIndex(long index) {
+        if (index < 0 || index > Integer.MAX_VALUE) {
+            throw new IndexOutOfBoundsException(
+                "Index out of bounds for SingleBuffer: " + index
+            );
+        }
+        return (int) index;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long capacity() {
+        return buffer.capacity();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long position() {
+        return buffer.position();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer position(long newPosition) {
+        buffer.position(checkIndex(newPosition));
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long limit() {
+        return buffer.limit();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer limit(long newLimit) {
+        buffer.limit(checkIndex(newLimit));
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte get() {
+        return buffer.get();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer get(byte[] dst) {
+        buffer.get(dst);
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public byte get(long index) {
+        return buffer.get(checkIndex(index));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public double getDouble() {
+        return buffer.getDouble();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public float getFloat() {
+        return buffer.getFloat();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer duplicate() {
+        return new SingleBuffer(this.buffer.duplicate());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long readFrom(FileChannel channel) throws IOException {
+        return channel.read(buffer);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String decode(CharsetDecoder decoder)
+        throws CharacterCodingException {
+        return decoder.decode(buffer).toString();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SingleBuffer asReadOnlyBuffer() {
+        return new SingleBuffer(buffer.asReadOnlyBuffer());
+    }
```

Comment:
I think we could drop this method and just call `asReadOnlyBuffer()` in the constructor of this and `MultiBuffer` as we will always want the underlying `ByteBuffer`s to be read only.